dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -522,10 +522,16 @@ void Spdp::process_location_updates_i(const DiscoveredParticipantIter& iter, con
     ACE_DEBUG((LM_DEBUG, "(%P|%t) DEBUG: Spdp::process_location_updates_i: %@ %C has %B location update(s) force_publish=%d reason=%C\n", this, LogGuid(iter->first).c_str(), location_updates.size(), force_publish, reason));
   }
 
+  const DCPS::TimeDuration leaseDuration = rtps_duration_to_time_duration(iter->second.pdata_.leaseDuration,
+                                             iter->second.pdata_.participantProxy.protocolVersion,
+                                             iter->second.pdata_.participantProxy.vendorId);
+
+  DCPS::ParticipantLocationBuiltinTopicData& location_data = iter->second.location_data_;
+  location_data.lease_duration = leaseDuration.to_dds_duration();
+
   bool published = false;
   for (DiscoveredParticipant::LocationUpdateList::const_iterator pos = location_updates.begin(),
-         limit = location_updates.end(); iter != participants_.end() && pos != limit; ++pos) {
-    DCPS::ParticipantLocationBuiltinTopicData& location_data = iter->second.location_data_;
+         limit = location_updates.end(); pos != limit; ++pos) {
```

Comment:
I think the `iter != participants_.end()` is probably just left-over code.  The invocation of the listener is scheduled on the Service Participant's thread which allows the lock to be held throughout.  Since the lock is held, the iterator should still be valid.